test: Enable the counter_basic_api test on mec172x board

hunterhu3000 opened the pull request:
1. enable timer5 on dts
2. Replace all BTMR_Type with struct btmr_regs, to make it compatiable
   on both mec 15 and 17
3. As MEC15 and MEC17 have different ways to access the register, use
   "#if defined" to access counter_cfg->girq_id on different board.
4. Fix a bug, when no device is found and the counter test is not
   acutaly run, it will still return PASS.

Test pass on mec172xevb_assy6906

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
@albertofloyd This PR is related to https://github.com/zephyrproject-rtos/zephyr/pull/45704. After it is merged, I'll fix the issue of https://github.com/zephyrproject-rtos/zephyr/issues/45647. Please review and comment. Thanks

jvasanth1 commented on dts/arm/microchip/mec172xnsz.dtsi:
```diff
@@ -421,7 +421,7 @@
 			prescaler = <0>;
-			status = "disabled";
+			status = "okay";
 		};
```
status okay should be in the board level file?

hunterhu3000 replied:
Agree. I'll move the code to board level dts file.

jvasanth1 commented on drivers/counter/counter_mchp_xec.c:
```diff
@@ -46,7 +46,7 @@ struct counter_xec_data {
 #define COUNTER_XEC_REG_BASE(_dev)			\
-	((BTMR_Type *)					\
+	((struct btmr_regs *)					\
 	 ((const struct counter_xec_config * const)	\
```
need to confirm this change works for MEC15xx as well.

hunterhu3000 replied:
The code works well on MEC15
BTMR_Type is defined like this in MEC15
```
typedef struct btmr_regs
{
        __IOM uint32_t CNT;     /*!< (@ 0x00000000) BTMR Count            */
        __IOM uint32_t PRLD;    /*!< (@ 0x00000004) BTMR Preload          */
        __IOM uint8_t  STS;     /*!< (@ 0x00000008) BTMR Status           */
        uint8_t RSVDC[3];
        __IOM uint8_t IEN;      /*!< (@ 0x0000000c) BTMR Interrupt Enable */
        uint8_t RSVDD[3];
        __IOM uint32_t CTRL;    /*!< (@ 0x00000010) BTMR Control          */
} BTMR_Type;
```
I tried to use typedef in soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h for MEC 17 soc/arm/microchip_mec/common/reg/mec_timers.h, but the check function says new typedef was not allowed. So I modify it in the driver code.

jvasanth1 commented on soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:
```diff
@@ -100,6 +103,18 @@
+
+/* 8 <= n <= 26 corresponding to GIRQ08, GIRQ09, ..., GIRQ26 */
+#define MCHP_GIRQ_SRC(n) REG32(MCHP_GIRQ_SRC_ADDR(n))
+
```
for MEC17xx we are not using macros to access GIRQ registers. We have functions in interrupt driver for this purpose.

hunterhu3000 replied:
Do you mean this?
MCHP_GIRQ_SRC ==> mchp_xec_ecia_girq_src_clr
MCHP_GIRQ_ENSET ==>mchp_xec_ecia_girq_src_en

Do you have an official version of counter_mchp_xec.c which can be used for both MEC15 and MEC17?
If so, could you please upload that file?

hunterhu3000 commented:
@jvasanth1 I have made a new patch, thanks to review.

MaureenHelm commented:
It's been a long time since CI ran on this PR. Please rebase to rerun CI

hunterhu3000 commented:
> It's been a long time since CI ran on this PR. Please rebase to rerun CI

OK, rebase is done

galak commented on drivers/counter/counter_mchp_xec.c:
```diff
@@ -243,14 +243,19 @@ static int counter_xec_set_top_value(const struct device *dev,
 
 	counter->STS = MCHP_BTMR_STS_ACTIVE;
+
+#if defined(CONFIG_BOARD_MEC172XEVB_ASSY6906)
```
Having board specific ifdef in driver code should be avoided.

albertofloyd replied:
@galak MEC152 uses a HAL and MEC172x does not the header definitions are very different.
Any suggestions @jvasanth1

galak replied:
Is this really board specific or SoC specific?

albertofloyd replied:
It's SoC specific maybe the change could be changed to CONFIG_SOC_SERIES_MEC172X

hunterhu3000 replied:
This is changed in the latest version

galak commented on drivers/counter/counter_mchp_xec.c:
```diff
@@ -294,7 +299,11 @@ static int counter_xec_init(const struct device *dev)
 	counter->PRLD = counter_cfg->info.max_top_value;
 	counter->CNT = counter_cfg->info.max_top_value;
 
+#if defined(CONFIG_BOARD_MEC172XEVB_ASSY6906)
```
Figure out how to do this so its not board specific

galak replied:
It feels like this is a SoC difference and not a board specific one.  If that is really the case, than I suggest just changing the ifdef to be based on SoC.

hunterhu3000 replied:
@galak Do you mean replace CONFIG_BOARD_MEC172XEVB_ASSY6906 with CONFIG_SOC_MEC172X_NSZ?
If yes, I have pushed a new patch and rebased. Thanks to review.

galak reviewed (dismissed):
replace board specific code in drivers

hunterhu3000 commented:
Rebased to the latest and fix the problem in https://github.com/zephyrproject-rtos/zephyr/issues/45647

MaureenHelm commented on soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:
```diff
@@ -23,4 +23,8 @@ config EEPROM_XEC
 	default y
 	depends on EEPROM
 
+config COUNTER_XEC
+	default y
+	depends on COUNTER
+
```
This can be removed following 0f1ed0f4a05

hunterhu3000 replied:
Yes, it can be removed.
The patch is updated. Thanks to review

MaureenHelm reviewed (changes requested):
Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit

MaureenHelm commented:
> Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit

This still needs to be done

hunterhu3000 committed:
tests: Enable the counter_basic_api test on mec172x board

1. enable timer5 on dts
2. Replace all BTMR_Type with struct btmr_regs, to make it compatiable
   on both mec 15 and 17
3. As MEC15 and MEC17 have different ways to access the register, use
   "#if defined" to access counter_cfg->girq_id on different board.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 committed:
tests: counter_basic_api will return pass even if no counter is found

In test_all_instances(), if ARRAY_SIZE(devices) is 0, the test is not
performed. At this time, the test result should be TC_FAIL, instead of
TC_PASS. A zassert is needed to filter this case.
Fixes: #45647

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
> > Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit
> 
> This still needs to be done

Done.
The CI fail looks irrelevant to the patches. The patches works well on mec172xevb. Thanks to check.

hunterhu3000 commented:
@albertofloyd could you please review and +1? Thanks

MaureenHelm commented:
@hunterhu3000 please rebase to pick up CI fixes
